GraphBase/src/graph/UnidirectedGraph.java

```diff
@@ -0,0 +1,37 @@
+package graph;
+
+public class UnidirectedGraph extends Graph {
+
+  public int addEdge(int fromId, int toId) {
+    this.addVertex(fromId);
+    this.addVertex(toId);
+    if (vertices.get(fromId).isAdjacent(toId)) {
+      return 0;//represents the edge has already existed.
+    }
+    Edge newEdge = new Edge(fromId, toId);
+    vertices.get(fromId).addFromAdjEdge(newEdge);
+    vertices.get(toId).addToAdjEdge(newEdge);
+    return 1; // normal return
+  }
+
+  public int addEdge(Vertex fromVertex, Vertex toVertex) {
+    return addEdge(fromVertex.getId(), toVertex.getId());
+  }
+
+  public int addEdge(Edge edge) {
+    return addEdge(edge.getFromId(), edge.getToId());
+  }
+
+  public Edge deleteEdge(int fromId, int toId) {
+    vertices.get(fromId).deleteAdjEdge(toId);
```

Comment:
Hey. Take care. You fromId and toId may not present in vertices. In this case, you will have NullPointerException.